ui/StatusQ/src/wallet/managetokenscontroller.cpp

```diff
@@ -0,0 +1,364 @@
+    m_settings.endGroup();
+    m_settings.sync();
+}
+
+void ManageTokensController::loadSettings()
+{
+    Q_ASSERT(!m_settingsKey.isEmpty());
+
+    m_settingsData.clear();
+
+    m_settings.beginGroup(QStringLiteral("ManageTokens-%1").arg(m_settingsKey));
+    const auto size = m_settings.beginReadArray(m_settingsKey);
+    for (auto i = 0; i < size; i++) {
+        m_settings.setArrayIndex(i);
+        const auto symbol = m_settings.value(QStringLiteral("symbol")).toString();
+        if (symbol.isEmpty()) {
+            qCWarning(manageTokens) << Q_FUNC_INFO << "Missing symbol while reading tokens settings";
+            continue;
+        }
+        const auto pos = m_settings.value(QStringLiteral("pos"), -1).toInt();
+        const auto visible = m_settings.value(QStringLiteral("visible"), true).toBool();
+        const auto groupId = m_settings.value(QStringLiteral("groupId")).toString();
+        m_settingsData.insert(symbol, {pos, visible, groupId});
+    }
+    m_settings.endArray();
+    m_settings.endGroup();
+}
+
+void ManageTokensController::revert()
+{
+    loadSettings();
+    parseSourceModel();
+}
+
+void ManageTokensController::classBegin()
+{
+    // empty on purpose
+}
+
+void ManageTokensController::componentComplete()
+{
+    loadSettings();
+}
+
+void ManageTokensController::setSourceModel(QAbstractItemModel* newSourceModel)
+{
+    if(!newSourceModel) return;
+    if(m_sourceModel == newSourceModel) return;
+    m_sourceModel = newSourceModel;
+
+    if (m_sourceModel && m_sourceModel->roleNames().isEmpty()) { // workaround for when a model has no roles and roles are added when the model is populated (ListModel)
+        // QTBUG-57971
+        connect(m_sourceModel, &QAbstractItemModel::rowsInserted, this, &ManageTokensController::parseSourceModel);
+    } else {
+        parseSourceModel();
+    }
+}
+
+void ManageTokensController::parseSourceModel()
+{
+    if (!m_sourceModel)
+        return;
+
+    disconnect(m_sourceModel, &QAbstractItemModel::rowsInserted, this, &ManageTokensController::parseSourceModel);
+
+#ifdef QT_DEBUG
+    QElapsedTimer t;
+    t.start();
+#endif
+
+    // clear all the models
+    for (auto model: m_allModels)
+        model->clear();
+    m_communityIds.clear();
+
+    // read and transform the original data
+    const auto newSize = m_sourceModel->rowCount();
+    qCInfo(manageTokens) << "!!! PARSING" << newSize << "TOKENS";
+    for (auto i = 0; i < newSize; i++) {
+        addItem(i);
+    }
+
+    // build community groups model
+    rebuildCommunityTokenGroupsModel();
+    reloadCommunityIds();
+    m_communityTokensModel->setCommunityIds(m_communityIds);
+
+    // (pre)sort
+    for (auto model: m_allModels) {
+        model->applySort();
+        model->saveCustomSortOrder();
+        model->setDirty(false);
+    }
+
+#ifdef QT_DEBUG
+    qCInfo(manageTokens) << "!!! PARSING SOURCE DATA TOOK" << t.nsecsElapsed()/1'000'000.f << "ms";
+#endif
+
+    emit sourceModelChanged();
```

Comment:
I'd remove this from `parseSourceModel`. It's not a parser job to emit it and we're also calling parseSourceModel from multiple places. Probably this is what's triggering the multiple connections on `rowsInserted` and other signals.

Reply:
Nah, it was the lambda (which I fixed); it's needed here because of the QTBUG mentioned above

Reply:
Yes, I've seen it. I was thinking that `sourceModelChanged` will be triggered unnecessarily on each call to `parseSourceModel`. And there's lots of code that will run on each `sourceModelChanged` event.
Now for the QTBUG you could have another method. Something like

```
void ManageTokensController::initialiseSourceModel()
{
         if (!m_sourceModel)
         return;

       disconnect(m_sourceModel, &QAbstractItemModel::rowsInserted, this, &ManageTokensController:: InitialiseSourceModel);
       parseSourceModel();
       emit sourceModelChanged();
}
```

Reply:
Yeah got it, and fixed by guarding the connection, check again the code pls